### PR TITLE
[gen4 planner] Make sure to not push down expressions when not possible

### PR DESCRIPTION
**Pull request by frouioui**
## Description

This PR fixes https://github.com/vitessio/vitess/issues/12540 by making sure to not push down to MySQL complex expressions that involved aggregations.

If we have a derived table and it's aggregating results for us, we can't use these results until the are fully aggregated, which happens on the vtgate level.

```sql
select dt.a,
       dt.b,
       (dt.a / dt.b)
from (select sum(a) as a,
             sum(b) as b
      from user) dt
```

In the query above, since we have to spread out the aggregation across all `user` shards, we can't push down the division between the two sums until after summing up the individual shard results.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/12540

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by systay — CHANGES_REQUESTED** [review]
I don't this is correct. The query we are trying to fix should not be using the `AggregateRandom` expression here in the first place

**Review by frouioui — COMMENTED** [review]
LGTM

**Review by arthurschreiber — APPROVED** [review]
❤️

File: go/vt/vtgate/executor_select_test.go [hunk 1]
```diff
@@ -3856,6 +3856,40 @@ func TestSelectAggregationData(t *testing.T) {
...
+		sbc.SetResults([]*sqltypes.Result{sqltypes.MakeTestResult(
+			sqltypes.MakeTestFields("a|b|c", "int64|int64|int64"),
+			"null|null|null",
+		)})
+	}
+
+	conns[0].SetResults([]*sqltypes.Result{sqltypes.MakeTestResult(
+		sqltypes.MakeTestFields("a|b|c", "int64|int64|int64"),
+		"10|1|10",
+	)})
```
Comment (harshit-gangal) [inline go/vt/vtgate/executor_select_test.go:3881]: the mock result should be changed to return only 2 columns from the shards.
`sum(a)` and `sum(b)`

Comment (frouioui) [reply go/vt/vtgate/executor_select_test.go]: Fixed via e9a203f00f

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**